File: static/js/game_objects.js
/**
 * Game Objects for Brick Breaker
 * This file defines the objects used in the game.
 */

// ======================
// Paddle Class
// ======================

// ======================
// Paddle Class
// ======================
class Paddle {
    constructor(canvasWidth, canvasHeight) {
        // Size
        this.width = 100;
        this.height = 20;
        
        // Position
        this.x = canvasWidth / 2 - this.width / 2;
        this.y = canvasHeight - 50;
        
        // Movement
        this.speed = 600; // pixels per second
        this.useMouseControl = true;
        
        // Canvas bounds
        this.canvasWidth = canvasWidth;
        
        // Laser powerup
        this.laserActive = false;
        this.laserTime = 0;
        this.laserCooldown = 0.3; // seconds between shots
        this.lastLaserTime = 0;
    }
    
    // Update paddle position based on input
    update(deltaTime, input) {
        if (!input) return;
        
        // Toggle control method if requested (M key)
        if (input.toggleControlPressed) {
            this.useMouseControl = !this.useMouseControl;
            console.log("Controls switched to:", this.useMouseControl ? "Mouse" : "Keyboard");
        }
        
        // Always respect forced keyboard mode (K key)
        if (input.forceKeyboardMode) {
            this.useMouseControl = false;
        }
        
        // Move paddle based on active control method
        if (this.useMouseControl && input.mouseX !== null) {
            // Mouse control
            this.x = input.mouseX - this.width / 2;
        } else {
            // Keyboard control
            if (input.leftPressed) {
                this.x -= this.speed * deltaTime;
            }
            if (input.rightPressed) {
                this.x += this.speed * deltaTime;
            }
        }
        
        // Keep paddle within canvas bounds
        this.x = Math.max(0, Math.min(this.canvasWidth - this.width, this.x));
        
        // Update laser timer
        if (this.laserActive) {
            this.laserTime -= deltaTime;
            if (this.laserTime <= 0) {
                this.laserActive = false;
            }
        }
    }
    
    // Activate laser powerup
    activateLaser(duration) {
        this.laserActive = true;
        this.laserTime = duration;
        this.lastLaserTime = 0; // reset cooldown
    }
    
    // Check if enough time has passed to fire laser
    canFireLaser() {
        const now = performance.now() / 1000;
        return now - this.lastLaserTime >= this.laserCooldown;
    }
    
    // Create laser objects
    fireLaser() {
        if (!this.laserActive || !this.canFireLaser()) {
            return [];
        }
        
        // Update last fire time
        this.lastLaserTime = performance.now() / 1000;
        
        // Create two lasers (one from each side of paddle)
        return [
            new Laser(this.x + 15, this.y - 10),
            new Laser(this.x + this.width - 15, this.y - 10)
        ];
    }
}

// ======================
// Ball Class
// ======================
class Ball {
    constructor(canvasWidth, canvasHeight, x, y) {
        // Size (for both display and collision)
        this.width = 15;
        this.height = 15;
        this.size = 15;
        
        // Position
        this.x = x !== undefined ? x : canvasWidth / 2 - this.size / 2;
        this.y = y !== undefined ? y : canvasHeight / 2;
        
        // Canvas bounds
        this.canvasWidth = canvasWidth;
        this.canvasHeight = canvasHeight;
        
        // Movement
        this.speedX = 0;
        this.speedY = 0;
        this.setInitialDirection();
        
        // State
        this.active = false;   // Is ball launched from paddle?
        this.thru = false;     // Can ball go through bricks?
    }
    
    // Set random initial direction
    setInitialDirection() {
        const angle = Math.random() * Math.PI / 4 + Math.PI / 4; // 45-90 degrees
        const speed = 450; // pixels per second
        
        this.speedX = speed * Math.cos(angle);
        this.speedY = -speed * Math.sin(angle); // Negative = up
    }
    
    // Update ball position
    update(deltaTime) {
        if (!this.active) return;
        
        // Move the ball
        this.x += this.speedX * deltaTime;
        this.y += this.speedY * deltaTime;
    }
    
    // Handle collision with walls
    handleWallCollision(canvasWidth, canvasHeight) {
        // Left and right walls
        if (this.x <= 0 || this.x + this.size >= canvasWidth) {
            this.speedX = -this.speedX;
            
            // Adjust position to prevent sticking
            if (this.x <= 0) {
                this.x = 0;
            } else {
                this.x = canvasWidth - this.size;
            }
        }
        
        // Top wall
        if (this.y <= 0) {
            this.speedY = -this.speedY;
            this.y = 0;
        }
    }
    
    // Position ball on paddle when not launched
    stickToPaddle(paddle) {
        this.x = paddle.x + paddle.width / 2 - this.size / 2;
        this.y = paddle.y - this.size;
    }
    
    // Handle collision with paddle
    handlePaddleCollision(paddle) {
        // Calculate where ball hit the paddle (0 to 1)
        const hitPosition = (this.x + this.size / 2 - paddle.x) / paddle.width;
        
        // Angle based on hit position (-60 to 60 degrees)
        const angle = (hitPosition - 0.5) * Math.PI / 3;
        
        // Get current speed
        const speed = Math.sqrt(this.speedX * this.speedX + this.speedY * this.speedY);
        
        // Set new direction with slight speed increase
        const newSpeed = Math.min(speed * 1.05, 800);
        this.speedX = newSpeed * Math.sin(angle);
        this.speedY = -Math.abs(newSpeed * Math.cos(angle)); // Force upward
        
        // Prevent ball from getting stuck in paddle
        this.y = paddle.y - this.size;
    }
    
    // Handle collision with brick
    handleBrickCollision(brick) {
        // Calculate collision depths
        const fromLeft = (this.x + this.size) - brick.x;
        const fromRight = (brick.x + brick.width) - this.x;
        const fromTop = (this.y + this.size) - brick.y;
        const fromBottom = (brick.y + brick.height) - this.y;
        
        // Find shallowest collision (smallest depth)
        const depths = [fromLeft, fromRight, fromTop, fromBottom];
        const minDepth = Math.min(...depths);
        
        // Bounce based on which side was hit
        if (minDepth === fromTop || minDepth === fromBottom) {
            // Hit top or bottom - reverse Y velocity
            this.speedY = -this.speedY;
            
            // Adjust position to prevent multiple hits
            if (minDepth === fromTop) {
                this.y = brick.y - this.size;
            } else {
                this.y = brick.y + brick.height;
            }
        } else {
            // Hit left or right - reverse X velocity
            this.speedX = -this.speedX;
            
            // Adjust position to prevent multiple hits
            if (minDepth === fromLeft) {
                this.x = brick.x - this.size;
            } else {
                this.x = brick.x + brick.width;
            }
        }
    }
}

// ======================
// Brick Class
// ======================
class Brick {
    constructor(x, y, strength = 1, powerupChance = 0.3) {
        // Size
        this.width = 75;
        this.height = 20;
        
        // Position
        this.x = x;
        this.y = y;
        
        // Strength (hits needed to break)
        this.strength = strength;
        this.maxStrength = strength;
        
        // Powerup properties will be set explicitly for editor bricks
        this.hasPowerup = false;
        this.powerupType = null;
        this.editorPlaced = false;
        
        // Only randomly assign powerups if not editor placed and not explicitly set
        if (!this.editorPlaced && Math.random() < powerupChance) {
            this.hasPowerup = true;
            this.powerupType = Math.floor(Math.random() * 8);
        }
    }
    
    // Handle being hit by ball or laser
    hit(ball = null) {
        // If hit by thru ball, break immediately
        if (ball && ball.thru) {
            this.strength = 0;
        } else {
            this.strength--;
        }
        
        // Return true if brick is broken
        return this.strength <= 0;
    }
    
    // Get color based on strength
    getColor() {
        const colors = [
            '#ff3232', // Red (strength 1)
            '#ffcc00', // Yellow (strength 2)
            '#32ff32', // Green (strength 3)
            '#3232ff'  // Blue (strength 4+)
        ];
        
        return colors[Math.min(this.strength, 4) - 1];
    }
}

// ======================
// Powerup Class
// ======================
class Powerup {
    constructor(x, y, type) {
        // Size
        this.width = 30;
        this.height = 30;
        this.size = 30;
        
        // Position
        this.x = x;
        this.y = y;
        
        // Properties
        this.type = type;       // Type of powerup
        this.speed = 150;       // Falling speed (pixels per second)
        this.collected = false; // Has it been collected?
        
        // Visual effects
        this.angle = 0;  // For rotation
    }
    
    // Update powerup position
    update(deltaTime) {
        if (this.collected) return;
        
        // Move downward
        this.y += this.speed * deltaTime;
        
        // Rotate for visual effect
        this.angle = (this.angle + 180 * deltaTime) % 360;
    }
    
    // Get color based on powerup type
    getColor() {
        const colors = [
            '#32ff32', // Expand (green)
            '#ff3232', // Shrink (red)
            '#00ccff', // Multi (cyan)
            '#3232ff', // Slow (blue)
            '#ff9900', // Fast (orange)
            '#ffcc00', // Laser (yellow)
            '#cc00ff', // Life (purple)
            '#ffcc00'  // Thru (yellow)
        ];
        
        return colors[this.type];
    }
    
    // Get symbol based on powerup type
    getSymbol() {
        const symbols = ['+', '-', 'M', 'S', 'F', 'L', '♥', 'T'];
        return symbols[this.type];
    }
}

// ======================
// Laser Class
// ======================
class Laser {
    constructor(x, y) {
        // Size
        this.width = 3;
        this.height = 15;
        
        // Position
        this.x = x;
        this.y = y;
        
        // Speed
        this.speed = 500; // pixels per second
    }
    
    // Update laser position
    update(deltaTime) {
        this.y -= this.speed * deltaTime;
    }
}

File: utils/game_objects.py
"""
Game Objects for Brick Breaker

This module defines the game objects used in the Brick Breaker game.
"""

import random
import math
import time

class Rect:
    """Simple rectangle for collision detection"""
    def __init__(self, x, y, width, height):
        self.x = x
        self.y = y
        self.width = width
        self.height = height
    
    def colliderect(self, other):
        """Check if this rectangle collides with another"""
        return (
            self.x < other.x + other.width and
            self.x + self.width > other.x and
            self.y < other.y + other.height and
            self.y + self.height > other.y
        )

class Paddle:
    """Player-controlled paddle that bounces the ball"""
    
    def __init__(self, screen_width, screen_height):
        self.original_width = 100  # Default paddle width
        self.width = self.original_width
        self.height = 20
        self.x = screen_width // 2 - self.width // 2
        self.y = screen_height - 50
        self.speed = 10
        self.rect = Rect(self.x, self.y, self.width, self.height)
        self.use_mouse = True
        self.laser_active = False
        self.laser_time = 0
        self.laser_cooldown = 0.5  # Seconds between laser shots
        self.last_laser_time = 0
        self.screen_width = screen_width
        self.move_left = False
        self.move_right = False
    
    def set_direction(self, left, right):
        """Set movement direction based on keyboard input"""
        self.move_left = left
        self.move_right = right
    
    def set_position_from_mouse(self, mouse_x):
        """Set paddle position based on mouse input"""
        self.x = mouse_x - self.width // 2
        # Keep paddle within screen bounds
        if self.x < 0:
            self.x = 0
        elif self.x > self.screen_width - self.width:
            self.x = self.screen_width - self.width
        self.rect.x = self.x
    
    def update(self, dt, input_data=None):
        """Update paddle position"""
        if self.use_mouse and input_data and 'mouse_x' in input_data:
            # Mouse control (handled in set_position_from_mouse)
            pass
        else:
            # Keyboard control
            if self.move_left:
                self.x -= self.speed
            if self.move_right:
                self.x += self.speed
            
            # Keep paddle within screen bounds
            if self.x < 0:
                self.x = 0
            elif self.x > self.screen_width - self.width:
                self.x = self.screen_width - self.width
        
        # Update rectangle position
        self.rect.x = self.x
        
        # Countdown laser time
        if self.laser_active:
            self.laser_time -= dt
            if self.laser_time <= 0:
                self.laser_active = False
    
    def shoot_laser(self, dt):
        """Create laser objects if cooldown allows"""
        current_time = time.time()
        if self.laser_active and current_time - self.last_laser_time >= self.laser_cooldown:
            self.last_laser_time = current_time
            return [
                Laser(self.x + 12, self.y - 10),
                Laser(self.x + self.width - 12, self.y - 10)
            ]
        return []

class Laser:
    """Laser projectile fired from the paddle"""
    
    def __init__(self, x, y):
        self.x = x
        self.y = y
        self.speed = 10
        self.width = 3
        self.height = 15
        self.rect = Rect(self.x, self.y, self.width, self.height)
    
    def update(self, dt):
        """Update laser position"""
        self.y -= self.speed
        self.rect.y = self.y

class Ball:
    """Ball that bounces around and breaks bricks"""
    
    def __init__(self, screen_width, screen_height, x=None, y=None, speed_x=None, speed_y=None):
        self.size = 15  # Ball diameter
        self.x = x if x is not None else screen_width // 2
        self.y = y if y is not None else screen_height // 2
        self.screen_width = screen_width
        self.screen_height = screen_height
        
        # If speed is not provided, give a random direction
        if speed_x is None or speed_y is None:
            angle = random.uniform(math.pi/4, 3*math.pi/4)  # Angle between 45 and 135 degrees
            speed = random.uniform(4, 5)
            self.speed_x = speed * math.cos(angle)
            self.speed_y = -speed * math.sin(angle)  # Negative for upward movement
        else:
            self.speed_x = speed_x
            self.speed_y = speed_y
            
        self.rect = Rect(self.x, self.y, self.size, self.size)
        self.active = False
        self.thru = False
    
    def update(self, dt):
        """Update ball position"""
        if self.active:
            # Move the ball
            self.x += self.speed_x
            self.y += self.speed_y
            
            # Update rectangle position
            self.rect.x = int(self.x)
            self.rect.y = int(self.y)
    
    def stick_to_paddle(self, paddle):
        """Position the ball on top of the paddle when not active"""
        self.x = paddle.x + paddle.width // 2 - self.size // 2
        self.y = paddle.y - self.size
        self.rect.x = int(self.x)
        self.rect.y = int(self.y)
    
    def handle_wall_collision(self, screen_width, screen_height):
        """Handle collisions with screen boundaries"""
        if self.x <= 0 or self.x >= screen_width - self.size:
            self.speed_x = -self.speed_x
        if self.y <= 0:
            self.speed_y = -self.speed_y
    
    def handle_paddle_collision(self, paddle):
        """Handle collision with the paddle"""
        # Calculate where on the paddle the ball hit
        hit_pos = (self.x + self.size // 2) - paddle.x
        hit_ratio = hit_pos / paddle.width
        
        # Angle the ball based on where it hit (between -60 and 60 degrees)
        angle = (hit_ratio - 0.5) * (2*math.pi/3)
        
        # Speed increases slightly with each paddle hit
        speed = math.sqrt(self.speed_x**2 + self.speed_y**2)
        speed = min(speed * 1.05, 12)  # Cap max speed
        
        # Set new velocity
        self.speed_x = speed * math.sin(angle)
        self.speed_y = -speed * math.cos(angle)
    
    def handle_brick_collision(self, brick):
        """Handle collision with a brick"""
        # Calculate which side of the brick was hit
        dx_entry = (self.rect.x + self.rect.width - brick.rect.x) if self.speed_x > 0 else (brick.rect.x + brick.rect.width - self.rect.x)
        dy_entry = (self.rect.y + self.rect.height - brick.rect.y) if self.speed_y > 0 else (brick.rect.y + brick.rect.height - self.rect.y)
        
        if dx_entry < dy_entry:
            # Horizontal collision
            self.speed_x = -self.speed_x
        else:
            # Vertical collision
            self.speed_y = -self.speed_y

class Brick:
    """Breakable brick that can contain a powerup"""
    
    def __init__(self, x, y, strength=1, powerup_chance=0.3):
        self.width = 75
        self.height = 20
        self.x = x
        self.y = y
        self.strength = strength  # Number of hits to break
        self.max_strength = strength  # Remember initial strength
        self.rect = Rect(self.x, self.y, self.width, self.height)
        self.broken = False
        
        # Initialize powerup properties but don't randomize
        # These will be set explicitly in level loading if needed
        self.has_powerup = False
        self.powerup_type = None
        self.editor_placed = False  # Flag for editor-placed bricks
        
        # Only randomly assign powerups for non-editor bricks if not explicitly set
        if not self.editor_placed and random.random() < powerup_chance:
            self.has_powerup = True
            self.powerup_type = random.randint(0, 7)
    
    def hit(self, ball=None):
        """Reduce brick strength when hit"""
        # If ball has thru ability, break brick immediately
        if ball and ball.thru:
            self.strength = 0
        else:
            self.strength -= 1
        
        if self.strength <= 0:
            self.broken = True
            return True  # Brick is broken
        return False  # Brick is damaged but not broken

class Powerup:
    """Collectable powerup that provides special abilities"""
    
    def __init__(self, x, y, powerup_type):
        self.x = x
        self.y = y
        self.type = powerup_type
        self.size = 30
        self.speed = 3
        self.rect = Rect(self.x, self.y, self.size, self.size)
        self.collected = False
        self.angle = 0  # For rotation effect in visual rendering
    
    def update(self, dt):
        """Update powerup position"""
        if not self.collected:
            self.y += self.speed
            self.rect.y = self.y
            self.angle = (self.angle + 2) % 360  # For visual rotation